pyrae/core.py

```diff
@@ -20,6 +20,12 @@ def __init__(self, html: str):
         self._soup: Optional[BeautifulSoup] = None
         self.html = html
 
+    def __getstate__(self):
+        return self.html
```

Comment:
Can you elaborate as of what you are trying to do here?
Normally __getstate__ returns a dict and __setstate__ modifies self.__dict__, so I am not comfortable with this proposal.

If you try to tell me the use case for pickling, I can add support for it.
What object instances are you trying to pickle??

Best if you can open an issue with a code example where I can consistently repro the problem you are seeing.

Thanks a lot for your suggestions and contributions.
I will try to be more active on this little project, but other projects are keeping me busy :)

Reply:
Right now pickling is not generally possible due to reference cycles in the HTML soup (I don't know if this is something to do with the implementation of BeautifulSoup or if it is just the nature of the HTML itself which might contain cycles). I've made an example in #5.

It is true that in many cases `__getstate__` and `__setstate__` operate with dictionaries, but it is by no means a requirement. These methods should just satisfy the requirement that `__setstate__` can accept what `__getstate__` returns to initialise a new bare object of the class. From the [docs]():

> `object.__getstate__()`
> Classes can further influence how their instances are pickled; if the class defines the method `__getstate__()`, it is called and the returned object is pickled as the contents for the instance, instead of the contents of the instance’s dictionary. If the `__getstate__()` method is absent, the instance’s `__dict__` is pickled as usual.
>
> `object.__setstate__(state)`
> Upon unpickling, if the class defines `__setstate__()`, it is called with the unpickled state. In that case, there is no requirement for the state object to be a dictionary. Otherwise, the pickled state must be a dictionary and its items are assigned to the new instance’s dictionary.

So a reasonable way to represent any `FromHTML` object in a picklable form is the HTML itself; to pickle an object we use the underlying HTML and to unpickle it we initialize a new object with the unpickled HTML.

Thanks for the project, it is a neat little utility!